### PR TITLE
feat: fallback to `NoopProvider` if OOM happens

### PR DESCRIPTION
**Pull request by kirillzyusko**
### Details

A follow up for https://github.com/Expensify/react-native-onyx/pull/439

This PR accumulates al changes from https://github.com/Expensify/react-native-onyx/pull/483 and https://github.com/Expensify/react-native-onyx/pull/475 and adds a usage of `MemoryOnlyProvider` in unit-tests.

Originally https://github.com/Expensify/react-native-onyx/pull/483 and https://github.com/Expensify/react-native-onyx/pull/475 were reverted - in this PR I added fixes:
- https://github.com/Expensify/react-native-onyx/pull/485/commits/c461b993af46703fa03006a3bef1750a5db592e6 <- fixes a crash when `this` is `undefined` (which prevented a synchronization between tabs).
- https://github.com/Expensify/react-native-onyx/pull/485/commits/42c42eac988f17760ada9518534f3f390965ed92 <- fixes infinite initialization

In this PR I added [1/3]:
- all calls to providers are managed via additional `Storage` layer (in future this layer will intercept errors and will substitute provider to InMemory);
- synchronization between tabs is not a responsibility of provider anymore - now it's handled in separate `InstanceSync` class/object - such approach allows us to avoid code duplication if we add `InMemory` provider on web;
- all providers are initialized lazily (all providers got a new `init` method);

In this PR I added [2/3]:
- `name` property to all providers (used in logger if we ran into OOM);
- added `NoopProvider` that is used as a fallback one when OOM happens;
- added `tryOrDegradePerformance` function that attempt to perform operation and in case of specific error substitutes the storage to a fallback one;

In this PR I added [3/3]:
- `MemoryOnlyProvider` implementation that stores all info in memory;
- re-worked mocks and used `MemoryOnlyProvider` there;

### Related Issues

https://github.com/Expensify/App/issues/29403

### Automated Tests

This PR changes the way we use the storage but the functionality of the library to the outer world is the same. Therefore no new tests were added.

### Manual Tests

Verify that no flows and functionality were broken by the changes. Check for console errors regarding Onyx.

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
@marcaaron @tgolen I assigned you as you were reviewer of the closed PR and because you're familiar with the Onyx library

**Comment by kirillzyusko** [comment]
It's not needed to review now, because it contains changes from part 2. Let's review & merge part 2 and after that I'll rebase this Pr and we can review it 👀

**Comment by kirillzyusko** [comment]
@danieldoglas @tgolen this PR is ready for review 😊

File: lib/storage/providers/MemoryOnlyProvider.ts
Comment (hayata-suenaga) [inline lib/storage/providers/MemoryOnlyProvider.ts:132]: why do we need to measure the database size?

File: lib/storage/providers/MemoryOnlyProvider.ts
Comment (tgolen) [inline lib/storage/providers/MemoryOnlyProvider.ts:95]: Should this resolve with `null` like `getItem()`?

File: lib/storage/providers/MemoryOnlyProvider.ts
Comment (tgolen) [inline lib/storage/providers/MemoryOnlyProvider.ts:116]:
```suggestion
    /*
     * Since this is an in-memory only provider, nothing special needs to happen here and it can just be a noop
     */
```

**Review by tgolen — CHANGES_REQUESTED** [review]

File: lib/storage/providers/MemoryOnlyProvider.ts
Comment (marcaaron) [reply lib/storage/providers/MemoryOnlyProvider.ts:132]: We added this for logging purposes. It only gets used when we encounter an error so we probably don't need this. In theory, the `MemoryOnlyProvider` should never fail or throw any errors (at least doesn't look like it can).

Comment (marcaaron) [reply lib/storage/providers/MemoryOnlyProvider.ts:132]: See: https://github.com/Expensify/react-native-onyx/blob/69b2d5973b4aef19ce783440df55dedb30d4efd2/lib/Onyx.js#L1037-L1045

And: https://github.com/Expensify/react-native-onyx/blob/69b2d5973b4aef19ce783440df55dedb30d4efd2/lib/Onyx.js#L1069-L1078

**Review by marcaaron — APPROVED** [review]
Changes LGTM - I agree with @hayata-suenaga that we do not need to measure the size of the store (or at least can't think of any useful reason to do it).

File: lib/storage/providers/MemoryOnlyProvider.ts
Comment (kirillzyusko) [reply lib/storage/providers/MemoryOnlyProvider.ts]: Corrected 🙌

**Comment by kirillzyusko** [comment]
> I agree with @hayata-suenaga that we do not need to measure the size of the store (or at least can't think of any useful reason to do it).

@marcaaron so do you want to return a mocked value from `getDatabaseSize` without `object-sizeof` usage?

File: lib/storage/providers/MemoryOnlyProvider.ts
Comment (tgolen) [reply lib/storage/providers/MemoryOnlyProvider.ts:95]: OK, great! That sounds fine as long as that's what's happening in the other providers. Also, I just checked and I can't find any uses of `.remove()` in Expensify/App and it looks like this method is only used internally.

**Review by tgolen — APPROVED** [review]

File: lib/storage/providers/MemoryOnlyProvider.ts
Comment (hayata-suenaga) [reply lib/storage/providers/MemoryOnlyProvider.ts:134]: ah I can see from the code that it's only used in a very rare case. Thank you so much for the explanation!

**Review by hayata-suenaga — DISMISSED** [review]
the code looks good to me

there is [a pending question](https://github.com/Expensify/react-native-onyx/pull/485#issuecomment-1978669078)

**Review by danieldoglas — APPROVED** [review]

**Comment by marcaaron** [comment]
>@marcaaron so do you want to return a mocked value from getDatabaseSize without object-sizeof usage?

yeah, it does not need to be accurate so I'd rather not add a dependency if we don't need one.

**Comment by kirillzyusko** [comment]
> yeah, it does not need to be accurate so I'd rather not add a dependency if we don't need one.

@marcaaron okay, make sense - I'll remove it 🙌

**Review by marcaaron — APPROVED** [review]

**Comment by kirillzyusko** [comment]
@tgolen @hayata-suenaga I think it's ready for second review. I wasn't able to reproduce https://github.com/Expensify/App/issues/37752 since I think it's only possible in staging env?

I checked a code that responsible for synchronization - found and fixed there an issue. Checked how synchronization code works and it seems like it works okay.

If you have any suggestions which other local tests I need to perform in order not to introduce regressions - please tell me, I'll be happy to perform these tests 😊

**Review by tgolen — APPROVED** [review]
Great, thanks for squashing this all together. To help protect against regressions, I suggest the following:

- Create a PR on E/App that points to this branch (ie. the last commit like I did [here](https://github.com/Expensify/App/pull/37804/files))
- Request from our @appdeployers team on Slack to create an Ad-Hoc build for the E/App PR
- Then we can request for QA to do a full regression test on that ad-hoc build

That will allow us to find any possible regressions before committing to merge this PR.

**Review by hayata-suenaga — DISMISSED** [review]
the test plan sounds good 👍

File: lib/storage/index.ts
Comment (marcaaron) [inline lib/storage/index.ts:44]: Should we also catch the error we are explicitly throwing here?

https://github.com/Expensify/react-native-onyx/blob/3df4fef6d78fbb8a986304e071403f7a9d8418c4/lib/storage/providers/IDBKeyValProvider.ts#L22

Comment (kirillzyusko) [reply lib/storage/index.ts:44]: @marcaaron good catch! Yes, since we can not open a connection to DB, so all calls to set/get/remove will throw exceptions.

So for me it makes sense to substitute a storage in order not to crash the app 👍

File: lib/storage/index.ts
Comment (kirillzyusko) [inline lib/storage/index.ts:25]: @tgolen @marcaaron also a question about this - do we want to fallback to `NoopProvider` (does nothing) or to `MemoryProvider` stores all data in RAM?

Comment (tgolen) [reply lib/storage/index.ts:25]: I think the memory provider would be better. That's what the log message assumes.

**Comment by tgolen** [comment]
I got word from QA this morning that they have begun testing the E/App PR. They will report any bugs they find on that PR and we can triage them from there.

**Comment by kirillzyusko** [comment]
@mountiny @tgolen just a friendly ping to check the status of this PR 👀 

I think QA have tested and didn't find anything new (only old issues that can be reproduced in any preview deploys - CORS issues).

Just curious whether it can this PR be merged or not?

**Review by tgolen — APPROVED** [review]

**Comment by tgolen** [comment]
OK, I took one last look at the code and it looks good to me. I'm glad QA hasn't found any more issues with this and sounds like it's ready to be released. 

@marcaaron @danieldoglas @hayata-suenaga I wouldn't mind having at least one more approval before merging this. Would one of you like the honors?

**Comment by hayata-suenaga** [comment]
thank you for the review, Tim! Unfortunately, I don't have time to go over the code again, so I'd wait for Daniel or Marc 😄

**Comment by marcaaron** [comment]
Sure, taking a look!

File: lib/storage/index.ts
Comment (marcaaron) [inline lib/storage/index.ts:23]: NAB, it would be useful to log the error message as well. Without knowing anything about _why_ we fell back to using the cache this alert is not really actionable.

**Review by marcaaron — APPROVED** [review]
I have a small note. I am not sure if it's worth blocking on and this PR has been open for quite a while so suggesting a follow up to address it instead.

File: lib/storage/index.ts
Comment (tgolen) [reply lib/storage/index.ts:23]: I always got tripped up on this too, but it is logged on the very next line 😁

Comment (tgolen) [reply lib/storage/index.ts:23]: Oh, but maybe you mean to send the error as part of `logAlert()`?

Comment (marcaaron) [reply lib/storage/index.ts:23]: Yep! That's what I meant. I do see the `console.error()` which is local (good) but might be desirable to have server logs. 

I backpedaled here because I'm not sure if we are going to be tracking the frequency of these errors or doing anything about them.

That said, I also wonder if logging alert will lead to issues that nobody can action... I am not sure if the JS logs create issues or not. Maybe they should and if so - this could probably be a `hmmm()`.

Comment (kirillzyusko) [reply lib/storage/index.ts:23]: @tgolen @marcaaron should I do these changes in follow up PR? 👀

Comment (tgolen) [reply lib/storage/index.ts:23]: Yeah, I think it would probably be good to change this to a `hmmm()` and also log the specific error. It would sure be nice to look at server logs to determine _IF_ the fallback is happening and also _WHY_.

Comment (kirillzyusko) [reply lib/storage/index.ts:23]: @tgolen make sense! I've created a follow up PR: https://github.com/Expensify/react-native-onyx/pull/524

**Pull request merged**
